File: lab9_4/q4.c
#include <stdio.h>
#include <math.h>

// returns first i where |a[i+1]-a[i]| <= tol, else -1
int firstStable(double a[], int n, double tol)
{
    int i;
    for (i = 0; i < n-1; i++)
        if (fabs(a[i+1]-a[i]) <= tol + 1e-9)   // difference small enough
            return i;
    return -1;   // never got within tolerance
}

int main()
{
    int n, i;
    double tol;
    printf("Enter n: ");
    scanf("%d", &n);
    double a[n];
    printf("Enter the readings: ");
    for (i = 0; i < n; i++)
        scanf("%lf", &a[i]);
    printf("Enter tolerance: ");
    scanf("%lf", &tol);
    printf("%d\n", firstStable(a, n, tol));
    return 0;
}

File: lab9_4/q4.py
import numpy as np

# returns first i where |a[i+1]-a[i]| <= tol, else -1
def firstStable(a, n, tol):
    for i in range(n-1):
        if abs(a[i+1]-a[i]) <= tol + 0.000000001:   # difference small enough
            return i
    return -1   # never got within tolerance

a = np.array(input("Enter the readings: ").split(), dtype=float)
n = len(a)
tol = float(input("Enter tolerance: "))
print(firstStable(a, n, tol))
